### PR TITLE
Revise SharedInformer godoc comments

### PR DESCRIPTION
**Pull request by matte21**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design

/kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

(1) Periodic resyncs trigger update notifications (see notes to reviewer for proof), previous comments stated that create notifications are triggered, 3ae8c86 fixes that (and a minor typo as well). 

(2) The delete of an API object with identifier X followed by a create of a different API object with same identifier X can be seen as a single update notification by a ResourceEventHandler; 
ff543dd tries to explicitly state that in the comments on SharedInformer. In the previous comments there was all the needed information to derive that fact but at least one user still had doubts (see [this](https://kubernetes.slack.com/archives/C0EG7JC6T/p1571054208147900) thread). We can also drop this, but if we choose to keep it notice that there could be other places where this information might be added (e.g. godoc comments on [ResourceEventHandler](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/controller.go#L180))

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

If this gets approved let me know before you apply LGTM so that I can squash commits.

Appendix: "proof" that periodic resyncs cause update notifications as opposed to create notifications. When a [periodic resync is done](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L267-L269), the reflector invokes its store Resync method. The store is a DeltaFIFO, its Resync method is [this](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L511). [Keys for all the objects in the SharedInformer's cache are retrieved](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L519) and for each one [syncKeyLocked is invoked](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L521), which [adds a Sync delta for the object](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L550) (there are minor exceptions, some deltas not enqueued but this is not relevant here). When the SharedInformer pops a Sync delta, it [queries its cache and delivers an update notification if the object in the delta is already in the cache](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L455-L462) ; this can only be true for deltas enqueued because of a period resync.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by matte21** [comment]
@yliaog this PR still lacks the approved label. Can you add it please?

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 1]
```diff
@@ -106,7 +106,18 @@ import (
...
 // For a given SharedInformer and client, the notifications are
 // delivered sequentially.  For a given SharedInformer, client, and
-// object ID, the notifications are delivered in order.
+// object ID, the notifications are delivered in order.  Because
+// `ObjectMeta.UID` has no role in identifying objects, if an object
+// O1 with ID (e.g. namespace and name) X and `ObjectMeta.UID` U1 in
+// the SharedInformer's local cache is deleted and shortly after
+// another relevant object O2 with ID X and `ObjectMeta.UID` U2 is
+// created a client has no guarantee of seeing O1's deletion and
+// O2's creation as two separate notifications, it might as well
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: I suggest replacing "might as well" with "could instead" or "might instead", because in common usage "it might as well X" (in a context where Y has been discussed) means that "it" could _choose_ to do X and get the same result as if "it" had chosen to do Y but with less effort.  But that is not the situation here, "it" is not choosing whether to do X or Y.

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 2]
```diff
@@ -106,7 +106,18 @@ import (
 // and index updates happen before such a prescribed notification.
 // For a given SharedInformer and client, the notifications are
 // delivered sequentially.  For a given SharedInformer, client, and
-// object ID, the notifications are delivered in order.
+// object ID, the notifications are delivered in order.  Because
+// `ObjectMeta.UID` has no role in identifying objects, if an object
+// O1 with ID (e.g. namespace and name) X and `ObjectMeta.UID` U1 in
+// the SharedInformer's local cache is deleted and shortly after
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: We have no meaningful characterization of "shortly" here, so I would drop it.  I would replace "shortly after" with "later.  But this is just a suggestion.

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 3]
```diff
@@ -106,7 +106,18 @@ import (
...
+// another relevant object O2 with ID X and `ObjectMeta.UID` U2 is
+// created a client has no guarantee of seeing O1's deletion and
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: I suggest inserting "then" between "created" and "a", to give a bit of a break and make the higher level structure of the sentence more explicit.

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 4]
```diff
@@ -106,7 +106,18 @@ import (
 // and index updates happen before such a prescribed notification.
 // For a given SharedInformer and client, the notifications are
 // delivered sequentially.  For a given SharedInformer, client, and
-// object ID, the notifications are delivered in order.
+// object ID, the notifications are delivered in order.  Because
+// `ObjectMeta.UID` has no role in identifying objects, if an object
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: This is a very long sentence and it has two commas.  When a sentence has two commas, they should set off a parenthetic remark --- but that is not what is happening here.  I suggest rewording, something like the following.

> Because `ObjectMeta.UID` has no role in identifying objects, it is possible that when (1) object O1 with ID (e.g. namespace and name) X and `ObjectMeta.UID` U1 in the SharedInformer's local cache is deleted and later (2) another object O2 with ID X and `ObjectMeta.UID` U2 is created the informer's clients are not notified of (1) and (2) but rather are notified only of an update from O1 to O2.

**Review by MikeSpreitzer — COMMENTED** [review]
This looks technically correct to me but I have some suggestions on the wording.

**Comment by MikeSpreitzer** [comment]
/assign @smarterclayton
(@yliaog is not an approver)

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go (hunk 4 above)
Comment (matte21) [reply staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: Reworded as you suggested in 21f59c9.

**Review by MikeSpreitzer — APPROVED** [review]
/LGTM

**Comment by MikeSpreitzer** [comment]
/assign @wojtek-t 
(also assigned to other work on shared informers, so maybe better than Clayton)

**Comment by wojtek-t** [comment]
/approve

Thanks!
